src/Elders.Cronus/Projections/ProjectionVersions.cs

```diff
@@ -41,55 +41,52 @@ public void ValidateVersion(ProjectionVersion version)
             }
         }
 
-        public ProjectionVersions WithoutTheGarbage()
+        public void Add(ProjectionVersion version)
         {
-            HashSet<ProjectionVersion> result = new HashSet<ProjectionVersion>();
+            ValidateVersion(version);
 
-            foreach (var version in versions)
+            if (version.Status == ProjectionStatus.NotPresent)
+                return;
+
+            if (version.Status != ProjectionStatus.Building)
             {
-                if (version.Status != ProjectionStatus.Building)
-                {
-                    var versionInBuild = this.Where(x => x == version.WithStatus(ProjectionStatus.Building)).SingleOrDefault(); // searches for building version for the version hash
-                    result.Remove(versionInBuild);
+                var versionInBuild = versions.Where(x => x == version.WithStatus(ProjectionStatus.Building)).SingleOrDefault(); // searches for building version for the version hash
+                versions.Remove(versionInBuild);
 
-                    if (version.Status != ProjectionStatus.Live)
-                        result.Add(version);
-                }
+                if (version.Status != ProjectionStatus.Live)
+                    versions.Add(version);
+            }
 
-                if (version.Status == ProjectionStatus.Building)
-                    result.Add(version);
+            if (version.Status == ProjectionStatus.Building)
+                versions.Add(version);
 
-                if (version.Status == ProjectionStatus.Live)
+            if (version.Status == ProjectionStatus.Live)
+            {
+                var canceled = versions.Where(x => x == version.WithStatus(ProjectionStatus.Canceled)).SingleOrDefault();
+                versions.Remove(canceled);
+
+                var timedout = versions.Where(x => x == version.WithStatus(ProjectionStatus.Timedout)).SingleOrDefault();
+                versions.Remove(timedout);
+
+                var currentLiveVer = GetLive();
+                if (currentLiveVer is null || currentLiveVer <= version)
                 {
-                    var canceled = this.Where(x => x == version.WithStatus(ProjectionStatus.Canceled)).SingleOrDefault();
-                    result.Remove(canceled);
-
-                    var timedout = this.Where(x => x == version.WithStatus(ProjectionStatus.Timedout)).SingleOrDefault();
-                    result.Remove(timedout);
-
-                    var currentLiveVer = GetLive();
-                    if (ReferenceEquals(null, currentLiveVer) || currentLiveVer <= version)
-                    {
-                        result.Remove(currentLiveVer);
-                        result.Add(version);
-                    }
+                    versions.Remove(currentLiveVer);
```

Comment:
Do we have any race conditions here?

Reply:
No, it is basically the old version of the method => https://gyazo.com/4b4cca50b87787eb6787833d3b36158f

Reply:
I understand that we have not had any issues then.